### PR TITLE
Update 0.9-stable release to work with ruby 2.6 & 2.7

### PR DESCRIPTION
**Pull request by Physium**
#### Purpose
0.9-stable version as of now does not work with ruby 2.6 and above ( excluding ruby 3 ) due to how const_get has become stricter in terms searching for constants.

Examples mentioned in this issue: #2443 

#### Changes
- Update Gemfile to handle higher rails version if ruby is > 2.5
- Patch serializer_for method to search serializers based on a list of potential naming conventions

Note: this is my first rare attempt at making a PR, pardon my newbie-ness and appreciate any feedback.

### CODE REVIEW

File: test/unit/active_model/serializer/has_many_polymorphic_test.rb [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 module ActiveModel
   class Serializer
-    class HasManyPolymorphicTest < ActiveModel::TestCase
+    class HasManyPolymorphicTest < Minitest::Test
```
Comment (bf4) [inline test/unit/active_model/serializer/has_many_polymorphic_test.rb:5]: what's the reason for this change in this PR? Is it required?

File: test/unit/active_model/array_serializer/serialization_test.rb [hunk 2]
```diff
@@ -49,6 +49,7 @@ def test_serializer_for_array_returns_appropriate_type
         def object.serializer_class; CustomSerializer; end
 
         assert_equal CustomSerializer, Serializer.serializer_for(object)
+        assert_equal CustomSerializer, Serializer.serializer_for('Custom')
```
Comment (bf4) [inline test/unit/active_model/array_serializer/serialization_test.rb:52]: this is the failing test, right?

is accepting a String a new behavior?

File: Gemfile [hunk 3]
```diff
@@ -27,6 +28,7 @@ else
   gem 'activemodel', gem_version
   gem 'actionpack', gem_version
   gem 'activerecord', gem_version, group: :test
+  gem 'rails-controller-testing' if version > '4.2'
```
Comment (bf4) [inline Gemfile:31]: do you mean >= 4.2?

File: lib/active_model/serializer.rb [hunk 4]
```diff
@@ -118,11 +122,21 @@ def strip_attribute(attr)
         attr
       end
 
+      def build_serializer_class_list(resource, options)
+        list = []
+        list << build_serializer_class(resource, options)
+        list << build_serializer_class(resource.class.name.demodulize, {})
```
Comment (bf4) [inline lib/active_model/serializer.rb:128]: lookinng at 0-10-stable lib/active_model_serializers/lookup_chain.rb for how we solved this when it came up there looks like the same solution  more or less, so that's good :)

```ruby
    module_function

    def namespace_for(klass)
      klass.name.deconstantize
    end

    def resource_class_name(klass)
      klass.name.demodulize
    end

    def serializer_from_resource_name(name)
      "#{name}Serializer"
    end

    def serializer_from(klass)
      name = resource_class_name(klass)
      serializer_from_resource_name(name)
    end
```

and that in the serializer we called `serializer_class = lookup_chain.map(&:safe_constantize).find { |x| x && x < ActiveModel::Serializer }`

File: test/unit/active_model/array_serializer/serialization_test.rb (hunk 2 above)
Comment (Physium) [reply test/unit/active_model/array_serializer/serialization_test.rb:52]: This test is fine. I added an extra test since i want to allow string instead of a whole class object. This somewhat relaxes the input for serializer_for. Not sure if its a good idea, but its one simple fix i can think off.

Upon upgrading to ruby 2.5 >, there are 17 tests that fails due to the const_get changes. I used thou failed test as a benchmark to validate if the fix work. Is that ok? Is there a need to include new test?

File: lib/active_model/serializer.rb (hunk 4 above)
Comment (Physium) [reply lib/active_model/serializer.rb:128]: I got some inspiration from v0.10 for sure, thou i am still a little unclear on how the find condition works out. How does `x < ActiveModel::Serializer` work? 

Also, is there a need for me to make any changes to replicate how things are done in v0.10?

**Comment by Physium** [comment]
@bf4 I added in a missing test case that I came across in my project. Found out that if you have namespace on your model and its association they do not really play well.

**Comment by bf4** [comment]
Remind me if I don't look at this soon

**Comment by Physium** [comment]
@bf4 apologies for bothering. is there a rough timeline for the review?

File: lib/active_model/serializer.rb (hunk 4 above)
Comment (bf4) [reply lib/active_model/serializer.rb:129]: `x < ActiveModel::Serializer` is just one of those weird-looking Ruby things for asking if a thing is a subclass of the thing on the right

Comment (bf4) [reply lib/active_model/serializer.rb:129]: I'm guessing the 0.10 solution would work here as well. Up to you if you want to make another PR after this is merged

File: test/test_helper.rb [hunk 5]
```diff
@@ -2,6 +2,8 @@
 require 'minitest/autorun'
 require 'active_model_serializers'
 require 'fixtures/poro'
+require 'rails-controller-testing'
```
Comment (bf4) [inline test/test_helper.rb:5]: this can always be required now?

File: test/unit/active_model/serializer/has_many_polymorphic_test.rb (hunk 1 above)
Comment (bf4) [reply test/unit/active_model/serializer/has_many_polymorphic_test.rb:5]: I guess I misread this as removing ActiveSupport::TestCase (and not ActiveModel::TestCase)

I recall in 0.10x we went with ActiveSupport::TestCase over Minitest::Test since it provided better compatibility across minitest versions

File: Gemfile (hunk 3 above)
Comment (Physium) [reply Gemfile:31]: not exactly sure and I cant seem to figure out where is this test metric at. Based on the gem file, rails-controller-testing is only added when version is set to 'master'.  

I dont think 0.9-stable has ever been configured to work with the offical release of rails 5 and above.

File: test/unit/active_model/serializer/has_many_polymorphic_test.rb (hunk 1 above)
Comment (Physium) [reply test/unit/active_model/serializer/has_many_polymorphic_test.rb:5]: Personally have never worked with minitest and i notice most of these tests were written back in 2013/14. I made the switch simply by referencing what other model test files were referencing. I thought it made more sense for them to me the same across the board.

File: test/test_helper.rb (hunk 5 above)
Comment (Physium) [reply test/test_helper.rb:5]: yes, previously this was automatically required pre rails 5. now its need to be required when needed for >= rails 5.

Comment (bf4) [reply test/test_helper.rb:5]: gotcha.  well, If these changes for you, I'm ok to merge them.

Comment (bf4) [reply test/test_helper.rb:5]: seems travis ci is dead and this branch wasn't running a CI. I ported some stuff over in https://github.com/rails-api/active_model_serializers/pull/2447 if you're interested in helping out (happy to let you take it over) and figure out why it's failing and if we care

Comment (Physium) [reply test/test_helper.rb:5]: I wouldn't mind taking a look when I have time. Would definitely love to rewrite the lookup to something similar to v0.10 as well. However it does feel like i'm the only one in the world right now still on v0.9 haha!

Just curious, will there still be a release once this gets merged? Or do i have to reference the stable branch for all latest patches?

Comment (Physium) [reply test/test_helper.rb:5]: appreciate your prompt response! looks good to me as well! I'll help to look into CI when possible.

**Comment by bf4** [comment]
released 0.9.9

**Comment by zavan** [comment]
FYI this 'patch' update wrecked havoc on my Rails v7.0.5/Ruby v3.2.2 app. A lot of namespaced serializer resolving problems.

**Comment by Physium** [comment]
> FYI this 'patch' update wrecked havoc on my Rails v7.0.5/Ruby v3.2.2 app. A lot of namespaced serializer resolving problems.

Are you able to share examples of the problem? Perhaps open an issue and i can take a look. I am surprise that v9 even works with rails 7 and ruby 3.

**Comment by bf4** [comment]
@zavan we can help with a detailed error report. Obviously nothing was supposed to break. If you're up to help with testing and or proposing a fix, that would be much appreciated.  I haven't used this gem myself in many years, so don't have a personal use case,  but am committed to helping maintain it.

**Comment by Physium** [comment]
@zavan also did 0.9.8 work with ur existing setup?

**Comment by zavan** [comment]
v0.9.8 does work, so I just downgraded to fix.

It's an old app that started on Rails v3 but I've been able to update everything except for a large API that uses AS.

I don't have time right now to go in depth and write some tests to replicate, but here's one example. Consider these serializers:

```ruby
class BusinessSerializer < ActiveModel::Serializer
  # This is the serializer for the `Business` model, not `Integration::MyIntegration::Business`.
  # The model has a `name` attribute.
  attributes :id, :name
end

class LocationSerializer < ActiveModel::Serializer
  attributes :id, :name

  # The model is an instance of `Integration::MyIntegration::Business`,
  # which doesn't have a serializer and does not have a `name` attribute.
  has_one :my_integration_business 
end
```

v0.9.8:
```ruby
irb(main):005:0> LocationSerializer.new(Location.first).as_json
=>
{
  "location" => {
    :id => 1,
    :name => "West Coast",
    :my_integration_business => {
      "id" => 2,
      "location_id" => 1,
      "business_id" => "465607d5-9cb1-4ffd-a796-b18e11c3decd", # This has nothing to do with the Business model. It's an external id.
      "created_at" => "2023-06-02T08:01:56.085-05:00",
      "updated_at" => "2023-06-02T09:03:52.946-05:00"
    },
  }
}
```

So as you can see, it didn't find a serializer for the `Integration::MyIntegration::Business` instance and instead just dumped all model attributes. The application relies on this behavior.

v0.9.9:
```ruby
irb(main):001:0> LocationSerializer.new(Location.first).as_json

=> /usr/local/lib/ruby/gems/3.2.0/gems/activemodel-7.0.5/lib/active_model/attribute_methods.rb:450:in `method_missing': undefined method `name' for #<Integration::MyIntegration::Business id: 2, location_id: 1, business_id: "465607d5-9cb1-4ffd-a796-b18e11c3decd", created_at: "2023-06-02 08:01:56.085837000 -0500", updated_at: "2023-06-02 09:03:52.946017000 -0500"> (NoMethodError)
```

AS tries to use the `BusinessSerializer` for the `Integration::MyIntegration::Business` model instance. Which breaks the application as the attributes don't match.

**Comment by Physium** [comment]
Thank you @zavan! I'll look into it when I have time. The change should not affect existing way of doing things, all I did was to add additional look up chains if the initial lookup did not match. This is puzzling.

**Comment by Physium** [comment]
I get the issue now. This is all boils down to how const_get became stricter in ruby 2.6 onwards and as a result the birth of this PR.

In 2.5 and below, the example you gave would have also return BusinessSerializer instead of nil. Hence, the purpose of this PR is to somewhat mimic that behaviour. It seems like you guys adhere to the stricter behaviour and thus this gem works with no issues even with ruby 3 (which is great of course).

Now, I personally have no idea what should be the ideal behaviour should be. Its only recently I've been tasked to work on a project that uses this gem and upgrading its ruby version has resulted in a ton of failures. 

I am fine with removing my release and leaving my "patch" within my personal fork. Eventually, given the status of this gem, is to also migrate out of AMS into something more actively maintained.

Thoughts @bf4 ?

**Comment by bf4** [comment]
Maybe split the difference with a feature flag option for old behavior vs new?

**Comment by Kartikvlog** [comment]
when i am upgrading rails 5 to rails 6.1then getting error ``` No serializer found for resource: #<Comments  _id``` we have association:  user has_many comments and comments belongs to user.

**Comment by Physium** [comment]
> when i am upgrading rails 5 to rails 6.1then getting error ` No serializer found for resource: #<Comments _id` we have association: user has_many comments and comments belongs to user.

@Kartikvlog are you able to share more details? What version of the serializer are you on before you upgraded? If anything can you revert the serializer to 0.9.8 to see if it works? Upgrading rails shouldn’t cause any issue at all.

File: lib/active_model/serializer.rb [hunk 6]
```diff
@@ -66,10 +66,14 @@ def serializer_for(resource, options = {})
             ArraySerializer
           end
         else
-          klass_name = build_serializer_class(resource, options)
-          Serializer.serializers_cache.fetch_or_store(klass_name) do
-            _const_get(klass_name)
-          end
+          search_list = build_serializer_class_list(resource, options)
```
Comment (bf4) [inline lib/active_model/serializer.rb:69]: Upon review,, I should have noticed these changes are out of scope for this pr and asked for them to be proposed separately.
